browser/brave_rewards/rewards_chrome_sync_client.cc

```diff
@@ -0,0 +1,216 @@
+#include "chrome/browser/gcm/instance_id/instance_id_profile_service_factory.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/signin/identity_manager_factory.h"
+#include "chrome/browser/sync/device_info_sync_client_impl.h"
+#include "chrome/common/channel_info.h"
+#include "components/gcm_driver/gcm_profile_service.h"
+#include "components/gcm_driver/instance_id/instance_id_profile_service.h"
+#include "components/invalidation/impl/fcm_invalidation_service.h"
+#include "components/invalidation/impl/fcm_network_handler.h"
+#include "components/invalidation/impl/per_user_topic_subscription_manager.h"
+#include "components/invalidation/impl/profile_identity_provider.h"
+#include "components/invalidation/impl/profile_invalidation_provider.h"
+#include "components/invalidation/public/identity_provider.h"
+#include "components/invalidation/public/invalidation_service.h"
+#include "components/sync/base/model_type.h"
+#include "components/sync/base/report_unrecoverable_error.h"
+#include "components/sync/driver/model_type_controller.h"
+#include "components/sync/driver/sync_service.h"
+#include "components/sync/driver/sync_user_settings.h"
+#include "components/sync/driver/syncable_service_based_model_type_controller.h"
+#include "components/sync/model/forwarding_model_type_controller_delegate.h"
+#include "components/sync/model/model_type_store_service_impl.h"
+#include "components/sync_device_info/device_info_prefs.h"
+#include "components/sync_device_info/device_info_sync_service_impl.h"
+#include "components/sync_device_info/local_device_info_provider_impl.h"
+#include "components/sync_preferences/pref_service_syncable.h"
+#include "content/public/browser/storage_partition.h"
+
+using browser_sync::DeviceInfoSyncClientImpl;
+using invalidation::FCMInvalidationService;
+using invalidation::FCMNetworkHandler;
+using invalidation::IdentityProvider;
+using invalidation::InvalidationService;
+using invalidation::PerUserTopicSubscriptionManager;
+using invalidation::ProfileIdentityProvider;
+using invalidation::ProfileInvalidationProvider;
+using sync_preferences::PrefServiceSyncable;
+using syncer::DeviceInfoPrefs;
+using syncer::DeviceInfoSyncService;
+using syncer::DeviceInfoSyncServiceImpl;
+using syncer::ForwardingModelTypeControllerDelegate;
+using syncer::LocalDeviceInfoProviderImpl;
+using syncer::ModelTypeController;
+using syncer::ModelTypeStoreService;
+using syncer::ModelTypeStoreServiceImpl;
+using syncer::SyncableServiceBasedModelTypeController;
+
+namespace {
+auto CreateScopedPrefServiceSyncable(
+    PrefServiceSyncable* pref_service_syncable) {
+  DCHECK(pref_service_syncable);
+  return pref_service_syncable->CreateScopedPrefService(nullptr, {});
+}
+
+auto CreateModelTypeStoreService(Profile* profile) {
+  DCHECK(profile);
+  return std::make_unique<ModelTypeStoreServiceImpl>(
+      profile->GetPath().AppendASCII("rewards_sync"));
+}
+
+auto CreateDeviceInfoSyncService(
```

Comment:
we definitely should not be doing this. I don't understand what's going on here, not only are

Reply:
You suggested this back then:
```
syncer::DataTypeController::TypeVector
RewardsSyncClient::CreateDataTypeControllers(syncer::SyncService* sync_service) {
  syncer::DataTypeController::TypeVector controllers;

  auto device_info_sync_client =
      std::make_unique<browser_sync::DeviceInfoSyncClientImpl>(profile);
  auto local_device_info_provider =
      std::make_unique<syncer::LocalDeviceInfoProviderImpl>(
          chrome::GetChannel(),
          chrome::GetVersionString(chrome::WithExtendedStable(false)),
          device_info_sync_client.get());

  auto device_prefs = std::make_unique<syncer::DeviceInfoPrefs>(
      GetPrefServiceSyncable(), base::DefaultClock::GetInstance());

  model_type_store_ = std::make_unique<syncer::ModelTypeStoreServiceImpl>(
      profile->GetPath().Append("rewards_sync"))

  device_info_sync_service_ = 
      std::make_unique<syncer::DeviceInfoSyncServiceImpl>(
          model_type_store_.get(),
          std::move(local_device_info_provider), std::move(device_prefs),
          std::move(device_info_sync_client),
          nullptr);

  controllers.push_back(std::make_unique<ModelTypeController>(
      syncer::DEVICE_INFO,
      /*delegate_for_full_sync_mode=*/
      std::make_unique<syncer::ForwardingModelTypeControllerDelegate>(
          device_info_sync_service_
              ->GetControllerDelegate()
              .get()),
      /*delegate_for_transport_mode=*/
      std::make_unique<syncer::ForwardingModelTypeControllerDelegate>(
          device_info_sync_service_
              ->GetControllerDelegate()
              .get())));

  const base::RepeatingClosure dump_stack =
        base::BindRepeating(&syncer::ReportUnrecoverableError, chrome::GetChannel());

  controllers.push_back(
      std::make_unique<SyncableServiceBasedModelTypeController>(
          syncer::PREFERENCES,
          model_type_store_->GetStoreFactory(),
          SyncableServiceForPrefs(GetPrefServiceSyncable(),
                                  syncer::PREFERENCES),
          dumpstack));


  return controllers;
}
```

`CreateDeviceInfoSyncService()` above is just a helper function (internal to `rewards_chrome_sync_client.cc`). It takes as parameter the `ModelTypeStoreServiceImpl` created by `CreateModelTypeStoreService()` (which is also just a helper function) and returns a `DeviceInfoSyncServiceImpl`, which is then used in `CreateDataTypeControllers()`.